File: hw5/loadISOLET.py
#!/usr/bin/env python3

"""
loadISOLET.py
basically the same structure as loadMNIST.py i think
"""

import numpy


def load_isolet(letters, typestr):
    """
    input:
        letters: an iterable within components in {1, 2, ..., 26}
        typestr: 'test' 'train' or 'all'
    output:
        voice (data matrix, float)
        lab (class labels), float, 2D

    load the ISOLET dataset, found at https://archive.ics.uci.edu/ml/datasets/ISOLET
    although i think my input files are a slightly different format than what's
    found there.

    Note: This takes about 4 seconds to load (regardless of what 'letters' is,
    it seems) when loading 'all', so not super necessary to pickle. still
    consider if the main loop runs long.
    
    """
    
    # again, hardcoded filenames are bad
    
    if typestr == 'test':
        X = numpy.loadtxt('data/ISOLET/isolet.test', dtype='d', delimiter=', ')

    elif typestr == 'train':
        X = numpy.loadtxt('data/ISOLET/isolet.training', dtype='d', delimiter=', ')
    elif typestr == 'all':
        X1 = numpy.loadtxt('data/ISOLET/isolet.training', dtype='d', delimiter=', ')
        X2 = numpy.loadtxt('data/ISOLET/isolet.test', dtype='d', delimiter=', ')
        X = numpy.vstack((X1, X2))
    
    voice, lab = X[:,:-1], X[:,-1]

    # make a boolean index array to pare down in letters isn't all.
    # can't use `in` directly, also can't be a list. fiik
    # this is fast enough though
    bi = numpy.array([label in letters for label in lab]) 

     
    
    # note lab has shape (7797,) (in the case of 'all')
    # so i'm gonna reshape to a 2d array.
    return voice[bi], lab[bi].reshape((-1,1))

File: matlab/util/loadISOLET.m
function [voice,lab] = loadISOLET(letters,type)

if( strcmp(type,'test') ),
    X = dlmread('isolet.test');
    T = X(:,end);
    X(:,end) = [];
elseif( strcmp(type,'train') ),
    X = dlmread('isolet.training');
    T = X(:,end);
    X(:,end) = [];
elseif( strcmp(type,'all') ),
    X1 = dlmread('isolet.training');
    T1 = X1(:,end);
    X1(:,end) = [];
    X2 = dlmread('isolet.test');
    T2 = X2(:,end);
    X2(:,end) = [];
    X = [X1;X2];
    T = [T1;T2];
else
    error('Please specify test,train or all');
end;

voice = X;
lab = T;

idx = bsxfun(@eq,lab,letters);
voice = voice( sum(idx,2) == 1 , :);
lab = lab( sum(idx,2) == 1 );